controller/network/network.go

```diff
@@ -321,35 +341,44 @@ func (network *Network) CreateSession(srcR *Router, clientId *identity.TokenId,
 		}
 
 		// 4a: Create Route Messages
-		rms, err := circuit.CreateRouteMessages(sessionId, terminator.GetAddress())
+		rms, err := circuit.CreateRouteMessages(attempt, sessionId, terminator.GetAddress())
 		if err != nil {
 			return nil, err
 		}
-
-		// 5: Route Egress
 		rms[len(rms)-1].Egress.PeerData = clientId.Data
-		peerData, err := sendRoute(circuit.Path[len(circuit.Path)-1], rms[len(rms)-1], network.options.TerminationTimeout)
+
+		// 5: Routing
+		logrus.Debugf("route attempt [#%d] for [s/%s]", attempt + 1, sessionId.Token)
+		peerData, cleanups, err := rs.route(attempt, circuit, rms, strategy, terminator)
+		for k, v := range cleanups {
+			allCleanups[k] = v
+		}
 		if err != nil {
-			strategy.NotifyEvent(xt.NewDialFailedEvent(terminator))
-			retryCount++
-			if retryCount > 3 {
-				return nil, err
-			} else {
+			logrus.Warnf("route attempt [#%d] for [s/%s] failed (%v)", attempt + 1, sessionId.Token, err)
+			attempt++
```

Comment:
redundant cast

Reply:
I don't see a cast on the lines you highlighted?

Reply:
Ah, the line _below_.